### PR TITLE
Add optimized methods to compute grouped reductions

### PR DESCRIPTION
**Pull request by nalimilan**
```julia
df = DataFrame(a = repeat(1:40000, outer=[20]),
               b = repeat(20000:-1:1, outer=[40]),
               c = randn(800000))
gd = groupby(df, :a)
using BenchmarkTools, Statistics

# Master
julia> @btime combine(gd, :c => sum);
  10.473 ms (439181 allocations: 17.39 MiB)

julia> @btime combine(gd, :c => sum, :b => mean);
  19.515 ms (719227 allocations: 26.24 MiB)

# This PR
julia> @btime combine(gd, :c => sum);
  1.578 ms (154 allocations: 946.95 KiB)

julia> @btime combine(gd, :c => sum, :b => mean);
  3.046 ms (186 allocations: 1.54 MiB)
```

### CODE REVIEW

File: src/groupeddataframe/grouping.jl [hunk 1]
```diff
@@ -387,6 +423,98 @@ function gen_fun(::Type{NT}, f2) where NT
...
+# Use reducedim_init to get a vector of the right type,
+# but trick it into using the expected length
+groupreduce_init(op, incol, gd) =
+    Base.reducedim_init(identity, op, view(incol, 1:length(gd)), 2)
+
+for (op, initf) in ((:max, :typemin), (:min, :typemax))
+    @eval begin
+        function groupreduce_init(::typeof($op), incol::AbstractVector{T}, gd) where T
+            outcol = similar(incol, length(gd))
+            # Comparison is possible only between CatValues from the same pool
```
Comment (nalimilan) [inline src/groupeddataframe/grouping.jl]: It's unfortunate that the standard approach doesn't work for categorical arrays. I've also realized something like `maximum(categorical([1 2; 3 4]), dims=1)` currently doesn't work for the same reason. Not sure what to do about this in general.

File: src/groupeddataframe/grouping.jl [hunk 2]
```diff
@@ -387,6 +423,98 @@ function gen_fun(::Type{NT}, f2) where NT
...
+    # Undo pool sharing done by groupreduce_init
+    if outcol isa CategoricalVector
+        U = Union{CategoricalArrays.leveltype(outcol),
+                  eltype(outcol) >: Missing ? Missing : Union{}}
+        outcol = CategoricalArray{U, 1}(outcol.refs, incol.pool)
+    end
+    if isconcretetype(T)
+        return outcol
+    else
+        copyto_widen!(Tables.allocatecolumn(typeof(first(outcol)), n), outcol)
```
Comment (nalimilan) [inline src/groupeddataframe/grouping.jl]: This is also not optimal, but it's required if we want to return the same element type as the non-optimized method. Hopefully it should be quite rare, since reducing over `Union{T,Missing}` columns will return `missing` if there is any missing value, which isn't very useful (though one can have a  `Union{T,Missing}` column without any missing value).

**Comment by bkamins** [comment]
A quick comment (I just glanced at the code). We could add methods for counting (this is pretty common), variance, standadrd deviation, covariance. Of course this could be added later, but maybe these cases will give raise to a more general API design (I am not saying it is not general enough - but it would be good to know it would support such cases).

File: src/dataframerow/utils.jl [hunk 3]
```diff
@@ -90,7 +90,7 @@ end
 function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
                          hash::Val = Val(true),
                          groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool, Bool}
+                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool}
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: comment above should be updated.

File: src/dataframerow/utils.jl [hunk 4]
```diff
@@ -149,16 +149,36 @@ function row_group_slots(cols::Tuple{CategoricalVector},
...
             refmap .+= 1
+            seen[1] = true
         else
             refmap[1] = ngroups
         end
         @inbounds for i in eachindex(groups)
-            groups[i] = refmap[col.refs[i]+1]
+            j = refmap[col.refs[i]+1]
+            groups[i] = j
+            seen[j] |= true
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: why `|=` and not simply `=`? Is it faster?

File: src/dataframerow/utils.jl [hunk 5]
```diff
@@ -149,16 +149,36 @@ function row_group_slots(cols::Tuple{CategoricalVector},
...
+                ngroups += seen[i]
+                refmap[i] = ngroups
+            end
+            @inbounds for i in eachindex(groups)
+                groups[i] = refmap[groups[i]]
+            end
         end
     end
-    return ngroups, UInt[], Int[], true, true
+    return ngroups, UInt[], Int[], true
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: maybe after squashing non-present groups `groups` vector should be resized just to make sure its length is appropriate (this should not be very expensive I think).

File: src/dataframerow/utils.jl [hunk 6]
```diff
@@ -130,13 +130,13 @@ function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
...
     end
-    return ngroups, rhashes, gslots, false, false
+    return ngroups, rhashes, gslots, false
 end
 
 function row_group_slots(cols::Tuple{CategoricalVector},
                          hash::Val{false},
                          groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool, Bool}
+                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool}
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: maybe rename this function to `row_group_slots!` to show that it mutates its argument?

File: src/dataframerow/utils.jl (hunk 5 above)
Comment (bkamins) [reply src/dataframerow/utils.jl]: Actually it seems we must do it as below you do:
```
stops = zeros(Int, ngroups)
@inbounds for g_ix in groups
    stops[g_ix] += 1
end
```

Can you disable this `@inbounds` in tests and do we have test with categorical vector that has an empty level?

Comment (bkamins) [reply src/dataframerow/utils.jl]: Also in lower code it is assumed that `length(groups) == ngroups`
(this means that maybe we should run tests of DataFrames.jl with globaly disabled `@inbounds` with `--check-bounds=yes` option)

File: src/groupeddataframe/grouping.jl [hunk 7]
```diff
@@ -223,9 +230,9 @@ function Base.map(f::Any, gd::GroupedDataFrame)
         resize!(starts, j-1)
         resize!(ends, j-1)
         ends[end] = length(idx)
-        return GroupedDataFrame(parent, gd.cols, collect(1:length(idx)), starts, ends)
+        return GroupedDataFrame(parent, gd.cols, idx, collect(1:length(idx)), starts, ends)
     else
-        return GroupedDataFrame(parent, gd.cols, Int[], Int[], Int[])
+        return GroupedDataFrame(parent, gd.cols, idx, Int[], Int[], Int[])
```
Comment (bkamins) [inline src/groupeddataframe/grouping.jl]: `idx` is undefined in this branch. Can you please also add a test that catches it?

File: src/groupeddataframe/grouping.jl [hunk 8]
```diff
@@ -366,9 +378,33 @@ _nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
...
-function gen_fun(::Type{NT}, f2) where NT
-    get_val(::Val{T}) where {T} = T
+struct Reduce{F, A}
+    f::F
+    adjust::A
+end
+Reduce(f) = Reduce(f, nothing)
+function (r::Reduce)(x)
+    v = reduce(r.f, x)
+    r.adjust === nothing ? v : r.adjust(v, length(x))
```
Comment (bkamins) [inline src/groupeddataframe/grouping.jl]: I would rather define it `r.adjust(v, x)` to allow any adjustment function. It is a bit more flexible then. Next, if we wanted even more generality (given my earlier comment what functions could be supported) - an even broader `Reduce` definition could be considered.

I would document and export `Reduce` so that users can write their custom reducers (e.g. geometric or harmonic mean).

File: src/groupeddataframe/grouping.jl [hunk 9]
```diff
@@ -366,9 +378,33 @@ _nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
...
+end
+
+check_reduction(f::Any) = f
+check_reduction(::typeof(sum)) = Reduce(Base.add_sum)
+check_reduction(::typeof(prod)) = Reduce(Base.mul_prod)
+check_reduction(::typeof(maximum)) = Reduce(max)
+check_reduction(::typeof(minimum)) = Reduce(min)
+check_reduction(::typeof(mean)) = Reduce(Base.add_sum, /)
```
Comment (bkamins) [inline src/groupeddataframe/grouping.jl]: If you wanted to follow my recommendation above here we would write:
```
check_reduction(::typeof(mean)) = Reduce(Base.add_sum, (s,x) -> s / length(x))
```

File: src/groupeddataframe/grouping.jl [hunk 10]
```diff
@@ -366,9 +378,33 @@ _nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
...
 
+for f in (:sum, :prod, :maximum, :minimum, :mean)
+    @eval begin
+        funname(::typeof(check_reduction($f))) = Symbol($f)
```
Comment (bkamins) [inline src/groupeddataframe/grouping.jl]: maybe a more specific method name than `funname` could be used?

**Comment by bkamins** [comment]
I have left first comments (+ a comment on https://github.com/JuliaData/CategoricalArrays.jl/issues/178)

**Comment by matthieugomez** [comment]
I have to say: it is amazing how one can implement all of these fast aggregate functions in one commit. data.table / dplyr are now a few years old but I don't think they optimize `std` / `var` yet.

File: src/groupeddataframe/grouping.jl (hunk 10 above)
Comment (nalimilan) [reply src/groupeddataframe/grouping.jl]: That's the name of the method we use elsewhere already. What's wrong with it?

File: src/dataframerow/utils.jl (hunk 4 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: I figured it could avoid a write when not needed, but apparently it's slightly slower:
```julia
function f(refmap, groups, refs)
    seen = fill(false, length(refmap))
    @inbounds for i in eachindex(groups)
        j = refmap[refs[i]+1]
        groups[i] = j
        seen[j] |= true
    end
    seen, groups
end

N = 100_000
ngroups = 10
groups = Vector{Int}(undef, N)
refs = rand(1:ngroups, N)
refmap = rand(1:ngroups+1, ngroups+1)

# With |=
julia> @btime f(refmap, groups, refs);
  163.944 μs (2 allocations: 128 bytes)

# With =
julia> @btime f(refmap, groups, refs);
  149.227 μs (2 allocations: 128 bytes)
```

File: src/dataframerow/utils.jl (hunk 5 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: The `groups` vector has one entry for each row, not one for each group. So AFAICT there's no problem here (and we already pass `--check-bounds=yes`).

File: src/groupeddataframe/grouping.jl (hunk 8 above)
Comment (nalimilan) [reply src/groupeddataframe/grouping.jl]: The problem is, the whole point of these optimizations is to avoid constructing the `x` vector. And even if we constructed it (using a view, it wouldn't be too costly, though probably significant), its length cannot be computed efficiently for `skipmissing` cases, while it's almost free to compute it on the fly. So for now I'd rather keep this internal, and discuss with e.g. SplitApplyCombine developers to see how this could be generalized. Special-casing `length` doesn't sound too bad since that's an essential information and it's very cheap to compute anyway.

File: src/groupeddataframe/grouping.jl (hunk 9 above)
Comment (nalimilan) [reply src/groupeddataframe/grouping.jl]: See reply above.

**Comment by nalimilan** [comment]
I have added commits which improve the logic and implement `var`, `std`, `first`, `last` and `length`. We don't support `count` at the moment as that function does something different (which is a bit annoying, as you need to pick an arbitrary column to get its length). When multiple pairs are passed, we now compute each function in turn, instead of computing all functions at the same time for a given group. This allows using an optimized method even if a non-optimized one is specified in another pair, makes the code simpler, and should be easier to parallelize.

> I have to say: it is amazing how one can implement all of these fast aggregate functions in one commit. data.table / dplyr are now a few years old and I don't think they optimize `std` / `var` yet.

Good point. AFAICT data.table uses a slow approach, allocating vectors for each group to compute the variance on it: https://github.com/Rdatatable/data.table/blob/b186b715837a069e06d1a49026ed501ae8a1e524/src/gsumm.c#L1151
I've mentioned this for addition to the H2O.ai benchmarks: https://github.com/h2oai/db-benchmark/issues/60

File: src/groupeddataframe/grouping.jl (hunk 10 above)
Comment (bkamins) [reply src/groupeddataframe/grouping.jl]: Why not use `nameof` directly here?

File: src/dataframerow/utils.jl (hunk 6 above)
Comment (bkamins) [reply src/dataframerow/utils.jl]: OK

File: src/groupeddataframe/grouping.jl (hunk 8 above)
Comment (bkamins) [reply src/groupeddataframe/grouping.jl]: OK

File: src/dataframerow/utils.jl (hunk 5 above)
Comment (bkamins) [reply src/dataframerow/utils.jl]: Thanks. Now I understand.

**Comment by bkamins** [comment]
> We don't support count at the moment as that function does something different

I see the problem, but do you see a workaround in the future?

File: src/groupeddataframe/grouping.jl (hunk 10 above)
Comment (nalimilan) [reply src/groupeddataframe/grouping.jl]: Because we want the name of the original function, which isn't stored anywhere in the struct. Also what `funname` currently does is return `function` for anonymous functions (which you recently improved unless I'm mistaken).

**Comment by nalimilan** [comment]
I'm not sure, I haven't thought a lot about it. We could allow `() => length` or similar to avoid the need to pick an arbitrary column.

File: src/groupeddataframe/grouping.jl [hunk 11]
```diff
@@ -366,62 +379,260 @@ _nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
...
+check_aggregate(::typeof(length)) = Aggregate(length)
+# SkipMissing does not support length
+
+for f in (:sum, :prod, :maximum, :minimum, :mean, :var, :std, :first, :last)
+    @eval begin
+        funname(::typeof(check_aggregate($f))) = Symbol($f)
+        funname(::typeof(check_aggregate($f∘skipmissing))) = :function
+    end
+end
+funname(::typeof(length)) = :length
```
Comment (bkamins) [inline src/groupeddataframe/grouping.jl]: is this needed? already `funname(length)` should return `:length`. right?

File: src/groupeddataframe/grouping.jl [hunk 12]
```diff
@@ -366,62 +379,260 @@ _nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
...
+                nfilled += 1
+                break
             end
-            if res isa Union{AbstractDataFrame, NamedTuple, DataFrameRow, AbstractMatrix}
-                throw(ArgumentError("a single value or vector result is required when passing " *
-                                    "a vector or tuple of functions (got $(typeof(res)))"))
+        end
+    end
+    if nfilled < length(outcol)
+        throw(ArgumentError("cannot compute maximum or minimum for groups with only missing values"))
```
Comment (bkamins) [inline src/groupeddataframe/grouping.jl]: this can be also called by `first` and `last`. right?

**Comment by bkamins** [comment]
Regarding `count` discussion:
* `count`, `any` and `all` functions could be supported for boolean vectors by `groupreduce` even given its current implementation. Right?
* then if we wanted to handle predicates with `count`, `any` and `all`, which would be useful I think then we could use signatures matching `Base.Fix1{typeof(count)}` etc. to catch and handle these cases (the current would be `by(df, :group_col, :col=>Base.Fix1(count, predicate)`, but maybe this could be shortened if `Fix1` stared being used in Base - or at least exported); this would be cleanest; but maybe it would be simpler for a user to pass a tuple `by(df, :group_col, :col=>(count, predicate)` as it should be also easy enough to match?

Of course we do not have to do this in this PR.

**Comment by nalimilan** [comment]
OK, I thought by `count` you mean counting rows in a group (i.e. `length`, but more intuitive). `count(pred, x)`, `any` and `all` shouldn't be too hard to support using the framework from this PR. The main issue is indeed the public API. I think we should investigate whether DataFramesMeta could automatically detect these functions (and things like `sum(skipmissing(:col))`) and transform them into something that can be optimized. Ideally that would be a generic API which can also be used directly, maybe living in SplitApplyCombine.

@andyferris @davidanthoff @piever Do you think it would be interesting/possible to move the optimized operations from this PR to a generic package like SplitApplyCombine, so that they can be shared across DataFrames, JuliaDB and Query?

File: test/grouping.jl [hunk 13]
```diff
@@ -606,6 +631,94 @@ module TestGrouping
         end
     end
 
+    struct TestType end
+    Base.isless(::TestType, ::Int) = true
+    Base.isless(::Int, ::TestType) = false
+    Base.isless(::TestType, ::TestType) = false
+
+    @testset "combine with aggregation functions" begin
+        Random.seed!(1)
```
Comment (bkamins) [inline test/grouping.jl]: We could run scenarios:
* `df = DataFrame(a = rand(1:5, 20), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20))`
* `df = DataFrame(a = categorical(rand(1:5, 20)), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20))`
* `df = DataFrame(a = categorical(rand(1:5, 20)), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20)); levels!(df.a, shuffle([levels(df.a); 10; 11]))`
* `df = DataFrame(a = categorical(rand(1:5, 20), ordered=true), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20))`
* `df = DataFrame(a = categorical(rand(1:5, 20), ordered=true), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20)); levels!(df.a, shuffle([levels(df.a); 10; 11]))`
e.g. 10 times to automatically improve the coverage.

Comment (nalimilan) [reply test/grouping.jl]: Unfortunately that's not so simple. Most operations don't work for categorical values. I could add some cases in the categorical-specific tests, but what do you want to test exactly? For example, unused levels shouldn't make any difference here, and AFAIK they are tested elsewhere already (where it's more relevant). But please point uncovered scenarios if they are missing.

Comment (bkamins) [reply test/grouping.jl]: My point is that it is a complex design so randomizing tests might catch something we have not thought of. The question is do we fully cover with tests:
* grouping on one column that is categorical ordered and unordered with missing level and non-missing level (these are 4 scenarios);
* the same but when grouping on multiple columns all categorical (possibly with mixed characteristics);
* the same but when grouping on multiple columns some categorical and some not-categorical;

If we have all these scenarios covered I am OK to leave it as is (or "reasonably covered" - i.e. we do not have to test things that you are 100% sure that work the same as the things that are tested).

But in general - that is why I have approved earlier because I think that it is only an extra precaution.

Comment (nalimilan) [reply test/grouping.jl]: OK. What I meant is that to avoid combinatorial explosion of tests, I've tried to test each feature only once. All variations regarding the *grouping* column are handled in tests for `groupby`, since they make no difference for the `combine` code -- and AFAICT empty and reordered levels are tested there already. I guess we could add more tests with variations on the *aggregated* column here, but it's hard to decide what's relevant and what's not...

Comment (bkamins) [reply test/grouping.jl]: OK - let us leave it as is. I think that aggregated column is less risky.

File: src/groupeddataframe/grouping.jl [hunk 14]
```diff
@@ -366,62 +381,259 @@ _nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
...
+end
+Reduce(f, condf=nothing) = Reduce(f, condf, nothing)
+
+check_aggregate(f::Any) = f
+check_aggregate(::typeof(sum)) = Reduce(Base.add_sum)
+check_aggregate(::typeof(prod)) = Reduce(Base.mul_prod)
+check_aggregate(::typeof(maximum)) = Reduce(max)
+check_aggregate(::typeof(minimum)) = Reduce(min)
+check_aggregate(::typeof(mean)) = Reduce(Base.add_sum, nothing, /)
+check_aggregate(::typeof(sum∘skipmissing)) = Reduce(Base.add_sum, !ismissing)
```
Comment (andyferris) [inline src/groupeddataframe/grouping.jl]: Cool... I hadn't thought to introspect the type of the composition in this way - but this is a really clear way of expressing this!

**Comment by andyferris** [comment]
> Do you think it would be interesting/possible to move the optimized operations from this PR to a generic package like SplitApplyCombine, so that they can be shared across DataFrames, JuliaDB and Query?

Sorry Milan, I'm not terribly familiar with some of the implementation of operations on DataFrames, so I'm having a bit of trouble following along. A grouped data frame is internally sorted by grouping key, rather than nested via `Dict`s?

I guess I'm wondering precisely what you are suggesting - there is already a generic `SplitApplyCombine.groupreduce` for arbitrary iterables which is optimized not to allocate or rearrange large amounts of memory (and I would love to have `groupvar` and `groupstd` and so-on as well, and combinations of everything working well with `skipmissing`). However, in the case of `SplitApplyCombine` it uses nesting more than the traditional SQL approach so the output is a `Dict` rather than a table with two columns.

> OK, I thought by count you mean counting rows in a group

I was also torn by `groupcount` vs `grouplength`.

**Comment by nalimilan** [comment]
> Sorry Milan, I'm not terribly familiar with some of the implementation of operations on DataFrames, so I'm having a bit of trouble following along. A grouped data frame is internally sorted by grouping key, rather than nested via `Dict`s?

Yes, it stores a reference to a parent `DataFrame`, and information about which row goes to which group (in different forms).

> I guess I'm wondering precisely what you are suggesting - there is already a generic `SplitApplyCombine.groupreduce` for arbitrary iterables which is optimized not to allocate or rearrange large amounts of memory (and I would love to have `groupvar` and `groupstd` and so-on as well, and combinations of everything working well with `skipmissing`). However, in the case of `SplitApplyCombine` it uses nesting more than the traditional SQL approach so the output is a `Dict` rather than a table with two columns.

OK, I thought there was a function to group values according to a vector of group indices in SplitApplyCombine, but apparently that's not the case. It turns out it's often faster to compute group memberships in a first step (since you can compute hashes using SIMD and efficient memory access patterns; and with `CategoricalArray` columns group indices are already available without hashing nor dicts), and then do the actual reduction in a second step (since you can again use a very efficient loop). At least that's the approach DataFrames uses.

If you're also interested in a method taking group indices (in addition to a function computing the group, as currently), some code could be moved from DataFrames to SplitApplyCombine. It's a small part of the code, but it can be tricky, in particular when determining the initial values (especially for `minimum` and `maximum` and when skipping missing values) and the output type by progressive widening. But the fact that SplitApplyCombine returns a dict and DataFrames a vector is problematic...

**Comment by andyferris** [comment]
> It turns out it's often faster to compute group memberships in a first step

Right. In *SplitApplyCombine* there is `groupinds` (and `groupview`) function but it does construct a `Dict`. Also it all works with *AcceleratedArrays* where secondary indices may store precomputed groups, that can be taken advantage of for a variety of operations including `groupreduce`. (The generic `groupreduce` operation I wrote has no O(input size) allocations so supports out-of-core streaming, which is the logic behind that as the default choice, though I'm happy to revisit any implemetation details).

I imagine there are a few more operations for different permutations of splitting data. For example, just speculating, if you already have the spliting indices then we could have some `split` and `splitreduce` functions...

> But the fact that SplitApplyCombine returns a dict and DataFrames a vector is problematic...

I just see this as: we need multiple generic functions for different types of nested/flattened inputs and outputs. Some will make sense for generic iteratables, others (like the SQL flavoured group operations) depend on concepts like columns (which so far `SplitApplyCombine` isn't aware of).

**Comment by davidanthoff** [comment]
Just on the Query integration question: I've been designing/thinking about for a while now an alternative backend that process things in a column fashion, rather than row wise. That had been the plan from the get go, but it is just slow going, and as a first step the row wise processing implementation right now is more generic (but less performant). The general idea is that this column based backend kicks in if a query is purely table based, and if the query is one that we can run in an efficient way in a columnar way. The design I'm playing with right now would also implement a bunch of whole query rewrites/optimizations like in a database. In any case, once we start to implement all of this, I would expect that we would want to use algorithms like the one in this PR here, so the work here is most likely going to be super helpful down the road for this Query project!

But, be warned, don't expect anything anytime soon. I'm both still tinkering with the design and assembling the team to work on this here at Berkeley.

**Comment by nalimilan** [comment]
OK, we don't have to do something right now, but if at some point you need code similar to this, we can try to share code across packages.

**Comment by nalimilan** [comment]
I've added tests for floating point corner cases. Good to go?

**Comment by bkamins** [comment]
Go ahead and merge it when you feel ready.

File: src/groupeddataframe/grouping.jl [hunk 15]
```diff
@@ -366,62 +381,259 @@ _nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
...
+function groupreduce!(res, f, op, condf, adjust,
+                      incol::AbstractVector{T}, gd::GroupedDataFrame) where T
+    n = length(gd)
+    if adjust !== nothing
+        counts = zeros(Int, n)
+    end
+    @inbounds for i in eachindex(incol, gd.groups)
+        gix = gd.groups[i]
+        x = incol[i]
+        if condf === nothing || condf(x)
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/groupeddataframe/grouping.jl]: I guess in the future LLVM should be able to generate them automatically, so I would leave this only as a reference.

**Comment by bkamins** [comment]
@nalimilan - I think it is good to go ahead and merge it (but please decide).
